File: backend/spec/entities/order/order_spec.rb
RSpec.describe Order do
  describe "#initialize" do
    context "when creating a new instance" do
      it "customer must be filled" do
        order = Order.new(customer)

        expect(order.customer).to be_truthy
      end
      it "items must be empty" do
        order = Order.new(customer)

        expect(order.items).to eq []
      end
      it "payment must be nil" do
        order = Order.new(customer)

        expect(order.payment).to be_falsey
      end
      it "closed_at must be nil" do
        order = Order.new(customer)

        expect(order.closed_at).to be_falsey
      end
    end
  end
  describe "#add_product" do
    context "when adding a new product" do
      it "must increment items length" do
        order = Order.new(customer)
        expected = order.items.length + 1;

        order.add_product(product, 5)

        expect(order.items.length).to eq expected
      end
    end
  end
  describe "#set_payment" do
    context "when closing the order" do
      it "payment must be filled" do
        order = Order.new(customer)

        order.set_payment(payment(order))

        expect(order.payment).to be_truthy
      end
      it "closed_at must be equal to paid_at" do
        order = Order.new(customer)
        payment = payment(order)

        order.set_payment(payment)

        expect(order.closed_at).to eq payment.paid_at
      end
    end
  end
  describe "#total_amount" do
    context "when the list is empty" do
      it "total_amount must be 0" do
        order = Order.new(customer)

        expect(order.total_amount).to eq 0
      end
    end
    context "when the list has items" do
      it "total_amount must be the sum of all items" do
        order = Order.new(customer)
        order.add_product(Product.new(name: 'first product', type: :membership, price: 1.00), 5)
        order.add_product(Product.new(name: 'second product', type: :membership, price: 10.00), 2)

        expect(order.total_amount).to eq 25
      end
    end
  end
end

def customer
  Customer.new(Address.new(zipcode: '45678-979'), Address.new(zipcode: '45678-979'), 'customer@email.com')
end

def payment(order)
  payment = Payment.new(order: order, payment_method: CreditCard.fetch_by_hashed('43567890-987654367'))
  payment.pay
  payment
end

def product
  Product.new(name: 'some product', type: :membership, price: 5.00)
end

File: backend/individual-assignment/typescript/src/order.spec.ts
import { Address } from "./address";
import { Customer } from "./customer";
import { Order } from "./order";
import { CreditCard, Payment } from "./payment";
import {
  Product,
  ProductType,
  BookProductProcessor,
  DigitalProductProcessor,
  PhysicalProductProcessor,
  MembershipProductProcessor,
} from "./product";

describe("order", () => {
  let order: Order;

  beforeEach(() => {
    jest.clearAllMocks();
    order = new Order(new Customer("Maria", "maria@gmail.com"), new Address());
  });

  const mockBookProductProcess = jest.fn();
  const mockDigitalProductProcess = jest.fn();
  const mockPhysicalProductProcess = jest.fn();
  const mockMembershipProductProcess = jest.fn();

  jest
    .spyOn(BookProductProcessor.prototype, "process")
    .mockImplementation(mockBookProductProcess);
  jest
    .spyOn(DigitalProductProcessor.prototype, "process")
    .mockImplementation(mockDigitalProductProcess);
  jest
    .spyOn(PhysicalProductProcessor.prototype, "process")
    .mockImplementation(mockPhysicalProductProcess);
  jest
    .spyOn(MembershipProductProcessor.prototype, "process")
    .mockImplementation(mockMembershipProductProcess);

  const shirt = new Product("Basic t-shirt", ProductType.PHYSICAL, 49.9);
  const max = new Product("Standard plan", ProductType.MEMBERSHIP, 29.9);
  const book = new Product(
    "The Hitchhiker's Guide to the Galaxy",
    ProductType.BOOK,
    120.0,
  );
  const music = new Product("Stairway to Heaven", ProductType.DIGITAL, 5.0);
  const card = new CreditCard("43567890-987654367");

  it("should add products to order", () => {
    order.addProduct(shirt, 2);

    expect(order.items).toHaveLength(1);
    expect(order.items[0].product).toEqual(shirt);
    expect(order.items[0].quantity).toEqual(2);

    order.addProduct(max, 1);

    expect(order.items).toHaveLength(2);
    expect(order.items[1].product).toEqual(max);
    expect(order.items[1].quantity).toEqual(1);
    expect(order.totalAmount).toEqual(129.7);

    order.addProduct(music, 1);

    expect(order.items).toHaveLength(3);
    expect(order.items[2].product).toEqual(music);
    expect(order.items[2].quantity).toEqual(1);
    expect(order.totalAmount).toEqual(134.7);

    order.pay(card);

    expect(mockBookProductProcess).not.toHaveBeenCalled();
    expect(mockDigitalProductProcess).toHaveBeenCalled();
    expect(mockPhysicalProductProcess).toHaveBeenCalled();
    expect(mockMembershipProductProcess).toHaveBeenCalled();
  });

  it("should not add the same product twice", () => {
    order.addProduct(book, 2);
    expect(() => order.addProduct(book, 1)).toThrow(
      "The product has already been added. Change the amount if you want more.",
    );
  });

  it("should process payment and close order", () => {
    order.addProduct(shirt, 2);

    expect(order.closedAt).toBeNull();

    order.pay(card);

    expect(order.closedAt).not.toBeNull();
    expect(order.payment).not.toBeNull();
    expect(order.payment).toBeInstanceOf(Payment);
    expect(order.payment?.paidAt).not.toBeNull();
    expect(order.payment?.authorizationNumber).toBeGreaterThan(0);
    expect(order.payment?.amount).toEqual(order.totalAmount);
    expect(order.payment?.invoice.billingAddress).toEqual(order.address);
    expect(order.payment?.invoice.shippingAddress).toEqual(order.address);
  });

  it("should not allow duplicate payments", () => {
    order.addProduct(shirt, 2);
    order.pay(card);
    expect(() => order.pay(card)).toThrow("The order has already been paid!");
  });

  it("should throw error when trying to pay empty order", () => {
    expect(() => order.pay(card)).toThrow("Empty order cannot be paid!");
  });

  it.each`
    product  | mockProcessFunction             | productType
    ${shirt} | ${mockPhysicalProductProcess}   | ${"PHYSICAL"}
    ${max}   | ${mockMembershipProductProcess} | ${"MEMBERSHIP"}
    ${book}  | ${mockBookProductProcess}       | ${"BOOK"}
    ${music} | ${mockDigitalProductProcess}    | ${"DIGITAL"}
  `(
    "should call correct processor when product type is $productType",
    ({ product, mockProcessFunction }) => {
      order.addProduct(product, 1);
      order.pay(card);

      expect(mockProcessFunction).toHaveBeenCalled();
    },
  );
});
